packages/shared/scope-hoisting/src/concat.js

```diff
@@ -123,8 +123,14 @@ export async function concat(bundle: Bundle, bundleGraph: BundleGraph) {
 }
 
 async function processAsset(bundle: Bundle, asset: Asset) {
-  let code = await asset.getCode();
-  let statements: Array<Statement> = parse(code, asset.filePath);
+  let statements: Array<Statement>;
+  if (asset.astGenerator && asset.astGenerator.type === 'babel') {
+    let ast = await asset.getAST();
+    statements = nullthrows(ast).program.program.body;
+  } else {
+    let code = await asset.getCode();
```

Comment:
I took a look at `@babel/parser` and there doesn't seem to be an option to pass along an existing sourcemap when parsing. Does Babel read and use a sourcemap comment when parsing?

Reply:
Apparently only `require("@babel/core").transform*` parses the comment (which wouldn't really work here either).
If we indeed use swc for env, this would be rather important.

Reply:
For now, I'm ok with merging as is. In order to do this properly, we'd need to parse the source map and code, traverse all AST nodes, and map their source location through the source map. At the moment, this case should never happen unless you replace the JSTransformer or put a transformer after it that returns a string instead of an AST. I think that should be fairly rare, so we can address this later if it comes up.